mesonbuild/environment.py

```diff
@@ -632,16 +633,22 @@ def detect_cs_compiler(self):
             return MonoCompiler(exelist, version)
         raise EnvironmentException('Unknown compiler "' + ' '.join(exelist) + '"')
 
-    def detect_vala_compiler(self):
-        exelist = ['valac']
-        try:
-            p, out = Popen_safe(exelist + ['--version'])[0:2]
-        except OSError:
-            raise EnvironmentException('Could not execute Vala compiler "%s"' % ' '.join(exelist))
-        version = search_version(out)
-        if 'Vala' in out:
-            return ValaCompiler(exelist, version)
-        raise EnvironmentException('Unknown compiler "' + ' '.join(exelist) + '"')
+    def detect_vala_compiler(self, want_cross):
+        popen_exceptions = {}
+        compilers, ccache, is_cross, exe_wrap = self._get_compilers('vala', 'VALAC', want_cross)
+        for compiler in compilers:
+            if isinstance(compiler, str):
+                compiler = [compiler]
+            arg = ['--version']
+            try:
+                p, out = Popen_safe(compiler + arg)[0:2]
+            except OSError as e:
+                popen_exceptions[' '.join(compiler + arg)] = e
+                continue
+            version = search_version(out)
+            if 'Vala' in out:
+                return ValaCompiler(compiler, version, is_cross)
```

Comment:
__[flake8]__
* [F405] 'ValaCompiler' may be undefined, or defined from star imports: .compilers

<!-- Comment by [SideCI](https://sideci.com) -->